utils/zerovec/src/zerovec/var/components.rs

```diff
@@ -0,0 +1,243 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use super::*;
+use std::marker::PhantomData;
+use std::{iter, mem};
+
+fn usizeify(x: PlainOldULE<8>) -> usize {
+    u64::from_unaligned(&x) as usize
+}
+
+/// A logical representation of the backing `&[u8]` buffer.
+///
+/// This is where the actual work involved in VarZeroVec happens
+///
+/// See [`SliceComponents::new()`] for information on the invariants involved
+pub struct SliceComponents<'a, T> {
+    /// The list of indices into the `things` slice
+    indices: &'a [PlainOldULE<8>],
```

Comment:
An important thing i want to note about the design here is that `indices[0]` will basically _always_ be 0. It's technically possible for it to be nonzero from correctly-crafted data, but that's not of any practical consequence . We could make the array smaller by one if we made this about _end_ index instead of _start_ index (start index becomes the previous element's end index), though in that case to avoid the `from_fn` hacks in `iter()` I'd prefer to include the end index for the last element instead so that `indices` still has the same length. It should not be a major change.